File: render_cgl.py
# Renders the cgl files in CGoL_out/ that are produced by conway.c into gifs

from PIL import Image
from PIL import ImageDraw
import math
import struct
import time

# First and last frames to include in animation. First frame is 0
start = 0
end = 54

# Prepend to input cgl files
in_prefix = "CGoL_out/"

# filename of the output without the extension
out_name = "Barge"

# The dimensions of the file output
out_width = 512
out_height = 512

# Duration to show each animation frame in milliseconds
duration = 100

# Whether to write the frame number in the upper-left corner
draw_frame_num = False

# Method of coloring live cells based on time. Each entry contains a number being the amount of time the cell has to be alive
# Before this color phases in fully, followed by the RGB color all in a 4-tuple. The colors are linearly interpolated.
live_coloring = [(0, 255, 255, 255)]

# The same as above, but for dead cells
dead_coloring = [(0, 255, 0, 0), (8, 196, 196, 0), (16, 160, 160, 80), (24, 0, 240, 0), (36, 0, 0, 240), (42, 196, 0, 196), (62, 255, 0, 0), (68, 0, 0, 0)]

# The coloring for cells that have always been dead
dflt_coloring = (0, 0, 0) 

start_time = time.time()

img_buf = []
for f in range(start, end+1):
	fin = open(in_prefix + str(f) + ".cgl", "rb")
	width = struct.unpack("<H", fin.read(2))[0]
	height = struct.unpack("<H", fin.read(2))[0]
	rawdata = fin.read(width*height)
	fin.close()
	
	img_buf.append(Image.new("RGB", (width, height)))
	data = []
	for i in rawdata:
		val = i#struct.unpack("B", i)[0]
		is_alive = val & 128
		surv_time = val % 128
		
		if val == 127:
			data.append(dflt_coloring)
		elif val < 127:
			if dead_coloring[0][0] >= surv_time:
				data.append(dead_coloring[0][1:])
				continue
			if dead_coloring[-1][0] <= surv_time:
				data.append(dead_coloring[-1][1:])
				continue
			for c in range(1, len(dead_coloring)):
				C = dead_coloring[c]
				L = dead_coloring[c-1]
				t = (surv_time - L[0]) / (C[0] - L[0])
				if C[0] > surv_time:
					data.append((round(L[1] + t * (C[1]-L[1])), round(L[2] + t * (C[2]-L[2])), round(L[3] + t * (C[3]-L[3]))))
					break
		else:
			if live_coloring[0][0] >= surv_time:
				data.append(live_coloring[0][1:])
				continue
			if live_coloring[-1][0] <= surv_time:
				data.append(live_coloring[-1][1:])
				continue
			for c in range(1, len(live_coloring)):
				C = live_coloring[c]
				L = live_coloring[c-1]
				t = (surv_time - L[0]) / (C[0] - L[0])
				if C[0] > surv_time:
					data.append((round(L[1] + t * (C[1]-L[1])), round(L[2] + t * (C[2]-L[2])), round(L[3] + t * (C[3]-L[3]))))
					break
			
	img_buf[f].putdata(data)
	if draw_frame_num:
		d = ImageDraw.Draw(img_buf[f])
		d.text((0,0), str(f), fill=(255, 255, 255))
	img_buf[f] = img_buf[f].resize((out_width, out_height), Image.NEAREST)

images_colored = time.time() - start_time

img_buf[0].save(out_name + ".gif", save_all=True, append_images=img_buf[1:], duration=duration, loop=0)

images_saved = time.time() - start_time - images_colored

total = images_colored + images_saved

m0 = math.floor(images_colored / 60)
s0 = round(images_colored % 60 * 100) / 100

m1 = math.floor(images_saved / 60)
s1 = round(images_saved % 60 * 100) / 100

m2 = math.floor(total / 60)
s2 = round(total % 60 * 100) / 100

print("Coloring Images: " + str(m0) + "m " + str(s0) + "s, Saving to gif: " + str(m1) + "m " + str(s1) + "s, Total: " + str(m2) + "m " + str(s2) + "s")
				

File: render_cgl.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#define STB_IMAGE_IMPLEMENTATION
#include "stb_image/stb_image.h"
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image/stb_image_write.h"

// Renders a sequence of cgl files which take the form "inp_prefix + frame_#.cgl", for instance "frames/glider_0.cgl" where inp_prefix is "frames/glider_" and frame_# is 0.
// The out_prefix works the same way for output files. For such files, you can use prefix to specify a path and out_type to specify the filetype, 0 for png and 1 for jpg.
// Use the width/height parameters to specify the dimensions of the output. The Game of Life Grid will be stretched to fit.
// Color mode can be set to basic (Black and white for dead and alive cells) or inverted basic with the values 0 and 1 respectively. For value 2, it will accept the values of the color parameters and use them
// to animate the colors. Mode 3 is the same but uses RGB instead of grayscale. Output image files will use the correct number of channels.
int render_cgl(int frames, char* inp_prefix, char* out_prefix, int out_width, int out_height, int color_mode, unsigned char** live_color, int lc_size, unsigned char** dead_color, int dc_size, unsigned char* dflt_color) {
	char fileinp_buffer[128];
	char fileout_buffer[128];
	
	unsigned short width;
	unsigned short height;
	int size; // Values of loaded cgl frame.
	
	unsigned char* rawdata;	// Holds loaded data
	unsigned char* raw_end;	// End of loaded data
	unsigned char* outdata; // Holds RGB
	unsigned char* img;		// Holds image object to be saved
	unsigned char* C;		// temp variable ("Color")
	unsigned char* L;		// temp variable ("Last Color")
	
	unsigned char val;		 // temp variable holds the cached value of a pointed-to unsigned char (just micro-optimization things)
	unsigned char is_alive;	 // whether the currently considered pixel is alive
	unsigned char surv_time; // The number of previous consecutive frames this pixel has had it's current value.
	
	// To hold the final output image with 3 channels.
	img = (unsigned char*) malloc(out_width * out_height * 3);
		
	// For image scaling
	double width_ratio, height_ratio;
	int tx, ty;
	
	double t;
	
	// Simple definitions for modes 0 and 1
	unsigned char LIVE, DEAD;
	LIVE = 255;
	DEAD = 0;
	if (color_mode == 1) {
		LIVE = 0;
		DEAD = 255;
		color_mode = 0;
	}
	
	FILE* fin;
	// Load each frame
	for (int f = 0; f < frames; f++) {
		// Generate filename
		snprintf(fileinp_buffer, 128, "%s%d.cgl", inp_prefix, f);
		
		// Load cgl file
		fin = fopen(fileinp_buffer, "rb");
		fseek(fin, 0, SEEK_SET);
		fread(&width, 2, 1, fin);
		fread(&height, 2, 1, fin);
		size = width * height;
		
		
		// malloc() must be called inside this loop as the grid size may differ between frames. TODO: Try realloc()
		rawdata = (unsigned char*) malloc(size);
		outdata = (unsigned char*) malloc(size*3); // Times 3 for three channels
		raw_end = (unsigned char*) rawdata+size;
		
		// Read remainder of file and close
		fread(rawdata, 1, size, fin);
		fclose(fin);
		
		//printf("%d, %d, %d\n", width, height, size);
		
		// For each pixel, convert to image color. At this point, the input data is 100% in the cgl format.
		// Iterator variables outside of the for loop because they cannot both correctly be declared inside the loop parameters.
		unsigned char* i;
		unsigned char* o;
		int w = 0;
		for (i = rawdata, o = outdata; i < raw_end; i++, o+=3) {
			val = *i;
			is_alive  = val & 128;
			surv_time = val & 127;
			
			//if (w == 32 && f == 0) {printf("\n"); w = 0;}
			w++;
			
			// Simple black and white coloring
			if (color_mode == 0) {
				if (is_alive) {
					o[0] = LIVE;
					o[1] = LIVE;
					o[2] = LIVE;
				} else {
					o[0] = DEAD;
					o[1] = DEAD;
					o[2] = DEAD;
				}
			// Grayscale lifetime coloring with linear interpolation
			} else if (color_mode == 2) {
				printf("TODO: Implement Grayscale coloring. Sorry.\n");
				return 1;
			// RGB lifetime coloring with linear interpolation
			} else if (color_mode == 3) {
				if (val == 127) {
					//if (f == 0) {printf("D");}
					o[0] = dflt_color[0];
					o[1] = dflt_color[1];
					o[2] = dflt_color[2];
				} else if (val < 127) {
					//if (f == 0) {printf("d");}
					// Test for values outside the given range, and use the appropriate entries
					if (dead_color[0][0] >= surv_time) {
						o[0] = dead_color[0][1];
						o[1] = dead_color[0][2];
						o[2] = dead_color[0][3];
					} else if (dead_color[dc_size-1][0] <= surv_time) {
						o[0] = dead_color[dc_size-1][1];
						o[1] = dead_color[dc_size-1][2];
						o[2] = dead_color[dc_size-1][3];
					// If the input value is in range, interpolate its color values from the two entries it lies between.
					} else {
						int c;
						for (c = 1; c < dc_size; c++) {
							C = dead_color[c];
							L = dead_color[c-1];
							if (C[0] > surv_time) {
								t = (surv_time - L[0]) / (double) (C[0] - L[0]);
								o[0] = round(L[1] + t * (C[1]-L[1]));
								o[1] = round(L[2] + t * (C[2]-L[2]));
								o[2] = round(L[3] + t * (C[3]-L[3]));
								break;
							}
						}
					}
				} else {
					// Test for values outside the given range, and use the appropriate entries
					if (live_color[0][0] >= surv_time) {
						o[0] = live_color[0][1];
						o[1] = live_color[0][2];
						o[2] = live_color[0][3];
					} else if (live_color[lc_size-1][0] <= surv_time) {
						o[0] = live_color[lc_size-1][1];
						o[1] = live_color[lc_size-1][2];
						o[2] = live_color[lc_size-1][3];
					// If the input value is in range, interpolate its color values from the two entries it lies between.
					} else {
						for (int c = 1; c < lc_size; c++) {
							C = live_color[c];
							L = live_color[c-1];
							if (C[0] > surv_time) {
								t = (surv_time - L[0]) / (double) (C[0] - L[0]);
								o[0] = round(L[1] + t * (C[1]-L[1]));
								o[1] = round(L[2] + t * (C[2]-L[2]));
								o[2] = round(L[3] + t * (C[3]-L[3]));
								break;
							}
						}
					}
				}
			}
		}
		
		// Calculate scaling ratios for this frame
		width_ratio = width / (double) out_width;
		height_ratio = height / (double) out_height;
		// Scale img in outdata to out_width, out_height and save.
		for (int y = 0; y < out_height; y++) {
			for (int x = 0; x < out_width; x++) {
				tx = floor(x * width_ratio);
				ty = floor(y * height_ratio);
				
				img[(x + out_width*y)*3] = outdata[(tx + width*ty)*3];
				img[(x + out_width*y)*3+1] = outdata[(tx + width*ty)*3+1];
				img[(x + out_width*y)*3+2] = outdata[(tx + width*ty)*3+2];
			}
		}
		
		// Save img
		snprintf(fileout_buffer, 128, "%s%d.png", out_prefix, f);
		stbi_write_png(fileout_buffer, out_width, out_height, 3, img, out_width*3);
		
		free(rawdata);
	}
	
	free(img);
}

// int render_cgl(int frames, char* inp_prefix, char* out_prefix, int out_width, int out_height, int duration, int color_mode, unsigned char** live_color, int lc_size, unsigned char** dead_color, int dc_size, unsigned char* dflt_color) {
int main() {
	unsigned char*  dflt_color = (unsigned char*) malloc(3);
	
	unsigned char** dead_color = (unsigned char**) malloc(4);
	for (int i = 0; i < 4; i++) {
		dead_color[i] = (unsigned char*) malloc(4);
	}
	
	unsigned char*  live_color = (unsigned char*) malloc(4);
	
	dflt_color[0] = 0; dflt_color[1] = 0; dflt_color[2] = 0;
	
	dead_color[0][0] = 0;  dead_color[0][1] = 255; dead_color[0][2] = 80;  dead_color[0][3] = 120;
	dead_color[1][0] = 8;  dead_color[1][1] = 20;  dead_color[1][2] = 180; dead_color[1][3] = 200;
	dead_color[2][0] = 48; dead_color[2][1] = 4;   dead_color[2][2] = 4;   dead_color[2][3] = 4;
	
	live_color[0] = 0; live_color[1] = 128; live_color[2] = 128; live_color[3] = 255;
	
	render_cgl(64, "CGoL_out/", "CGoL_render/", 128, 128, 0, &live_color, 1, dead_color, 3, dflt_color);
	
	for (int i = 0; i < 4; i++) {
		free(dead_color[i]);
	}
	free(dead_color);
}